project_template/_base/project_name/apps/common/models/core.py

```diff
@@ -0,0 +1,54 @@
+from uuid import uuid4
+
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+
+class CoreQuerySet(models.QuerySet):
+
+    def active(self):
+        return self.filter(is_active=True)
+
+    def inactive(self):
+        return self.filter(is_active=False)
+
+
+class CoreManager(models.Manager):
+
+    def get_queryset(self):
+        return CoreQuerySet(self.model, using=self._db)
+
+    def active(self):
+        return self.get_queryset().active()
+
+    def inactive(self):
+        return self.get_queryset().inactive()
+
+
+class CoreModel(models.Model):
+
+    uuid = models.UUIDField(primary_key=True, default=uuid4, editable=False)
+    created = models.DateTimeField(auto_now_add=True, verbose_name=_('created'))
+    updated = models.DateTimeField(auto_now=True, verbose_name=_('updated'))
+    is_active = models.BooleanField(default=True, db_index=True)
+
+    objects = CoreManager()
+
+    class Meta:
+        abstract = True
+
+    def __str__(self):
+        return str(self.pk)
+
+    def __repr__(self):
+        return f'<{self.__class__.__name__} {self.pk}>'
```

Comment:
Maybe I'm wrong, but f-string is the slowest method of strings interpolation. So that using it for such simple cases is not optimal.

Reply:
I like new f-style string formatting, because it has the best readability.
If we'll talk about speed: formatting with `%` is the fastest way, but it's ugly as shit.

For `__repr__` in base model class it's not mission critical, so I've changed it to `format`.

Reply:
@romanosipenko `f` string are pre-parsed and stored in efficient bytecode for faster execution
Agree with @alexryabtsev that `%` looks oldy

Reply:
@sonerayberk thank you.

@romanosipenko it's not just faster, like Soner mentioned - it's a fastest method.

```
>>> import timeit
>>>
>>> timeit.timeit('val="string"; res="I am a %s" % val', number=100_000)
0.020300239004427567
>>> timeit.timeit('val="string"; res="I am a {}".format(val)', number=100_000)
0.03386643200065009
>>> timeit.timeit('val="string"; res="I am a {0}".format(val)', number=100_000)
0.041360289993463084
>>> timeit.timeit('val="string"; res="I am a {val}".format(val=val)', number=100_000)
0.05256030001328327
>>> timeit.timeit('val="string"; res=f"I am a {val}"', number=100_000)
0.010275076987454668
```

```
>>> import timeit
>>>
>>> timeit.timeit('val1="one"; val2="two"; res="I am number %s. And I am number %s." % (val1, val2)', number=100_000)
0.032395688991528004
>>> timeit.timeit('val1="one"; val2="two"; res="I am number {}. And I am number {}.".format(val1, val2)', number=100_000)
0.051123018987709656
>>> timeit.timeit('val1="one"; val2="two"; res="I am number {0}. And I am number {1}.".format(val1, val2)', number=100_000)
0.05416180597967468
>>> timeit.timeit('val1="one"; val2="two"; res="I am number {val1}. And I am number {val2}.".format(val1=val1, val2=val2)', number=100_000)
0.0822641899867449
>>> timeit.timeit('val1="one"; val2="two"; res=f"I am number {val1}. And I am number {val2}."', number=100_000)
0.015430753992404789
```

So I've returned f-style formatting back.

Reply:
That was in 3. It was slow. Initially, there is the discussion for some internal details https://stackoverflow.com/questions/37365311/why-are-literal-formatted-strings-so-slow-in-python-3-6-alpha-now-fixed-in-3-6

Useful warning/note post from Armin http://lucumr.pocoo.org/2016/12/29/careful-with-str-format/

If it is the fastest method, there are no reasons to reject your proposal, but I feel something that I can't explain :)

Reply:
I know - it looks like PHP, but if not overuse it it's more handful than `.format(...)`.
In all latest projects I use f-style formatting and totally happy with it.